react/src/layout/navbar.js

```diff
@@ -0,0 +1,204 @@
+    getChildLayouts = (config) => {
+        let { modal } = config;
+        return modal.items.map(layoutConfig => {
+            return {
+                layout: this.layoutsWrapper.getLayoutForConfig(layoutConfig),
+                config: layoutConfig
+            }
+        });
+    }
+
+    getFormDependencies = (config) => {
+        return this.getChildLayouts(config).map(child => child.layout.getFormDependencies(child.config)).flat();
+    }
+
+    component = (props) => {
+        let { config } = props;
+
+        let models = useContext(ContextModelsWrapper);
+        let title = useInterpolatedString(models, config.title);
+        let modalTitle = useInterpolatedString(models, config.modal.title);
+
+        let [modalShown, updateModalShown] = useState(false);
+
+        let formController = useContext(ContextRelativeFormController);
+        let simulationInfoPromise = useContext(ContextSimulationInfoPromise);
+        let modelsWrapper = useContext(ContextModelsWrapper);
+        
+        let store = useStore();
+
+        let _cancel = () => {
+            updateModalShown(false);
+            formController.cancelChanges();
+        }
+
+        let _submit = () => {
+            formController.saveToModels();
+            simulationInfoPromise.then(simulationInfo => {
+                modelsWrapper.saveToServer(simulationInfo, store.getState());
+            })
+        }
+
+        let children = this.getChildLayouts(config).map((child, idx) => {
+            let LayoutElement = child.layout.component;
+            return <LayoutElement key={idx} config={child.config}></LayoutElement>
+        })
+
+        let isDirty = formController.isFormStateDirty();
+        let isValid = formController.isFormStateValid();
+        let actionButtons = <ViewPanelActionButtons canSave={isValid} onSave={_submit} onCancel={_cancel}></ViewPanelActionButtons>
+
+        let { Portal: NavbarPortal, portalRef } = usePortal({
+            bindTo: document && document.getElementById(NavbarContainerId)
+        })
+
+        if(portalRef && portalRef.current) {
+            portalRef.current.classList.add("float-left");
+            portalRef.current.classList.add("col");
+        }
+
+        let { icon } = config;
+        let iconElement = undefined;
+        if(icon && icon != "") {
+            iconElement = <FontAwesomeIcon fixedWidth icon={Icon[icon]}></FontAwesomeIcon>;
+        }
+
+        // TODO fix button cursor on hover
+        return (
+            <>
+                <NavbarPortal>
+                    <Col>
+                        <div onClick={() => updateModalShown(true)} variant="secondary">
+                            <span>{title}<a className="ms-2">{iconElement}</a></span>
+                        </div>
+                    </Col>
+                </NavbarPortal>
+
+                <Modal show={modalShown} onHide={() => _cancel()} size="lg">
+                    <Modal.Header className="lead bg-info bg-opacity-25">
```

Comment:
I think classNames should come from somewhere to avoid errors where the names aren't found. bg-opacity-25 is something that should be defined elsewhere. `lead` is also something about formatting and should be defined in the style.